ai-safety-debate/confusion_matrix.py

```diff
@@ -0,0 +1,156 @@
+
+    index_of_lying_agent = (index_of_true_agent + 1)%2
+    
+    # confusion_matrix_counter[label][lying_agent_label] = number of times the lying agent won 
+    # by arguing for the 'lying_agent_label' while the true label being "label"
+    confusion_matrix_counter = np.array([[0 for i in range(10)] for i in range(10)], dtype='f')
+
+    # counts the types of labels of samples
+    labels_frequency = [0 for i in range(10)]
+
+    # play a debate game for each sample and lying label
+    for sample_id in range(nmbr_of_samples):
+
+        print("Sample", sample_id)
+        sample = judge.eval_data[sample_id].flatten()
+
+        label = judge.eval_labels[sample_id]
+
+        labels_frequency[label] += 1
+        
+        for lying_agent_label in range(10):
+
+            if lying_agent_label == label:
+                continue
+
+            agent_lie = DebateAgent(precommit_label=lying_agent_label, agentStrength=rollouts)
+            agent_truth = DebateAgent(precommit_label=label, agentStrength=rollouts)
+            assert index_of_true_agent in [0, 1]
+
+            player_description = DebatePlayers(
+                agent_lie, agent_truth, index_of_lying_agent, our_name="liar", opp_name="truth"
+            )
+
+            debate = Debate(
+                player_description.agents,
+                judge,
+                N_to_mask,
+                sample,
+                debug=False,
+                changing_sides=changing_sides,
+            )
+            utility = debate.play()
+
+            liar_won = player_description.increment_our_wincount(utility)
+
+            confusion_matrix_counter[label][lying_agent_label] += liar_won
+
+    print("Unnormalized confusion matrix")
+    print(confusion_matrix_counter)
+
+    print("Frequency of labels")
+    print(labels_frequency)
+
+    # normalize according to frequencies of the labels and multiply by 100 
+    normalized_confusion_matrix = confusion_matrix_counter
+    for label in range(10):
+        if labels_frequency[label] != 0:
+            normalized_confusion_matrix[label] = confusion_matrix_counter[label] * 1/labels_frequency[label]*100
+    print("Normalized confusion matrix")
+    print(normalized_confusion_matrix)
+
+    # adjust figure size
+    if dataset == "mnist":
+        plt.figure(figsize=(6,6))
+    elif dataset == "fashion":
+        plt.figure(figsize=(8,7))
+    else:
+        plt.figure(figsize=(10,5))
+    plt.matshow(normalized_confusion_matrix, cmap=plt.cm.jet, fignum=1)
+
+    cbar = plt.colorbar()
+    cbar.set_label('Liar win percentage')
+
+    # TODO liar label at the top
+    plt.xlabel('Liar label')
+    plt.ylabel('True label')
+
+
+    if dataset == "mnist":
+        plt.xticks(np.arange(10))
+        plt.yticks(np.arange(10))
+    elif dataset == "fashion":
+        plt.xticks(np.arange(10), fashion_labels, rotation='vertical')
+        plt.yticks(np.arange(10), fashion_labels)
+
+    plt.savefig('confusion_matrix.png')
```

Comment:
I am wondering if it makes sense to instead / additionally save the numpy arrays (maybe even the unnormalized ones). Then we could precompute them and play around with the figure afterwards. Also, this gives us the option to look into the actual numbers, if something is not clear in the figure. What do you think?